theano/sandbox/multinomial.py

```diff
@@ -192,6 +192,181 @@ def perform(self, node, ins, outs):
                             z[0][n, m] += 1
                             break
 
+class WeightedSelectionFromUniform(Op):
+    """
+    Converts samples from a uniform into sample from a multinomial.
+
+    """
+
+    __props__ = ("odtype",)
+
+    def __init__(self, odtype):
+        self.odtype = odtype
+
+    def __str__(self):
+        return '%s{%s}' % (self.__class__.__name__, self.odtype)
+
+    def __setstate__(self, dct):
+        self.__dict__.update(dct)
+
+    def make_node(self, pvals, unis, n=1):
+        pvals = T.as_tensor_variable(pvals)
+        unis = T.as_tensor_variable(unis)
+        if pvals.ndim != 2:
+            raise NotImplementedError('pvals ndim should be 2', pvals.ndim)
+        if unis.ndim != 1:
+            raise NotImplementedError('unis ndim should be 1', unis.ndim)
+        if self.odtype == 'auto':
+            odtype = pvals.dtype
+        else:
+            odtype = self.odtype
+        out = T.tensor(dtype=odtype, broadcastable=pvals.type.broadcastable)
+        return Apply(self, [pvals, unis, as_scalar(n)], [out])
+
+    def grad(self, ins, outgrads):
+        pvals, unis, n = ins
+        (gz,) = outgrads
+        return [T.zeros_like(x) for x in ins]
```

Comment:
- gradient wrt `n` should be a `DisconnectedType()` (for `MultinomialFromUniform` as well)
- for `pvals` and `unis`, I think there are arguments for both. I think that the Op itself should return `zeros`, as it is purely deterministic, and flat almost everywhere.
- However, I think it would be better if the wrapper method `multinomial` or `weighted_selection` would also add "undefined_grad" nodes, so that the whole process of "getting a sample from a multinomial" would not be derivable. We are not doing that yet in MRG, so I think it is fine for now.